torch/distributed/_sharded_tensor/api.py

```diff
@@ -113,6 +141,68 @@ def __init__(
         else:
             raise ValueError(f'Unsupported sharding_spec: {self._sharding_spec}')
 
+        with _sharded_tensor_lock:
+            global _sharded_tensor_current_id, _sharded_tensor_map
+            self._sharded_tensor_id = _sharded_tensor_current_id
+            _sharded_tensor_map[self._sharded_tensor_id] = self
+            _sharded_tensor_current_id += 1
+
+        # Initialize RPC if available.
+        if rpc._is_current_rpc_agent_set():
+            self._init_rpc()
+
+    def __del__(self):
+        # Clean up the global map.
+        with _sharded_tensor_lock:
+            global _sharded_tensor_current_id, _sharded_tensor_map
+            if self._sharded_tensor_id in _sharded_tensor_map:
+                _sharded_tensor_map.pop(self._sharded_tensor_id)
+
+    def _init_rpc(self):
+        self._remote_shards = []
+
+        # Gather all the sharded tensor ids.
+        world_size = dist.get_world_size(self._process_group)
+        worker_infos = rpc._get_current_rpc_agent().get_worker_infos()
+        rank_to_name = {}
+        name_to_rank = {}
+
+        for worker_info in worker_infos:
+            rank_to_name[worker_info.id] = worker_info.name
+            name_to_rank[worker_info.name] = worker_info.id
+
+        rpc_workers = set()
+        for rank in range(world_size):
+            if self._process_group == distributed_c10d._get_default_group():
+                global_rank = rank
+            else:
+                global_rank = distributed_c10d._get_global_rank(self._process_group, rank)
+            rpc_workers.add(rank_to_name[global_rank])
+
+        all_tensor_ids = rpc.api._all_gather(self._sharded_tensor_id, rpc_workers)
+
+        # Share the local shards to the entire world.
+        futs = []
+        for rank in range(world_size):
+            # Skip self.
+            if rank == dist.get_rank(self._process_group):
+                continue
+
+            if self._process_group == distributed_c10d._get_default_group():
+                global_rank = rank
+            else:
+                global_rank = distributed_c10d._get_global_rank(self._process_group, rank)
+
+            if len(self.local_shards()) != 0:
+                rrefs = [rpc.RRef(shard) for shard in self.local_shards()]
+                fut = rpc.rpc_async(global_rank, _register_remote_shards, args=(all_tensor_ids[rank_to_name[global_rank]], rrefs))
+                futs.append(fut)
+
+        torch.futures.wait_all(futs)
```

Comment:
Just want to confirm: the execution of each future here should be very fast right? I don't see any data movement, so adding each RRef should be pretty lightweight. Otherwise, we may need to consider deferring `wait_all` or only `wait` on individual remote shard on demand.

Reply:
Yes, there is no data movement here. We are only sending RRefs to other processes so they know about all shards.